File: java/SwapNodesInPairs.java
public class SwapNodesInPairs {
    public ListNode swapPairs(ListNode head) {
        ListNode temp = new ListNode(0);
        temp.next = head;
        head = temp;
        ListNode ptr1 = head;
        ListNode ptr2 = head.next;
        while (ptr2 != null && ptr2.next != null) {
            ListNode token = ptr2.next.next;
            ptr1.next = ptr2.next;
            ptr2.next.next = ptr2;
            ptr2.next = token;
            ptr1 = ptr2;
            ptr2 = ptr2.next;
        }
        return head.next;
    }
}

File: cc/swap_nodes_in_pairs.cc
#include "base_include.h"

class SwapNodesInPairs {
public:
    ListNode *swapPairs(ListNode *head) {
        ListNode *token = new ListNode(0);
        token->next = head;
        head = token;
        ListNode *curr = head;
        ListNode *post = head->next;
        ListNode *temp = NULL;
        while (post != NULL && post->next != NULL) {
            temp = post->next->next;
            curr->next = post->next;
            post->next = temp;
            curr->next->next = post;
            curr = post;
            post = post->next;
        }
        head = head->next;
        delete token;
        return head;
    }
};
